trie/committer.go

```diff
@@ -0,0 +1,284 @@
+const LeafChanSize = 200
+
+// Leaf represents a trie leaf value
+type Leaf struct {
+	size   int         // size of the rlp data (estimate)
+	hash   common.Hash // hash of rlp data
+	node   node        // the node to commit
+	vnodes bool        // set to true if the node (possibly) contains a valueNode
+}
+
+// committer is a type used for the trie Commit operation. A committer has some
+// internal preallocated temp space, and also a callback that is invoked when
+// leaves are committed. The leafs are passed through the `leafCh`,  to allow
+// some level of paralellism.
+// By 'some level' of parallelism, it's still the case that all leaves will be
+// processed sequentially - onleaf will never be called in parallel or out of order.
+type committer struct {
+	tmp sliceBuffer
+	sha keccakState
+
+	onleaf LeafCallback
+	leafCh chan *Leaf
+}
+
+// committers live in a global db.
+var committerPool = sync.Pool{
+	New: func() interface{} {
+		return &committer{
+			tmp: make(sliceBuffer, 0, 550), // cap is as large as a full fullNode.
+			sha: sha3.NewLegacyKeccak256().(keccakState),
+		}
+	},
+}
+
+// newCommitter creates a new committer or picks one from the pool, and
+// initializes the leafCh, if needed.
+// In case no onleaf-callback is provided, the committer does not
+// use a channel-based commit, but inlined.
+// Typically, the account trie is committed with a channel-based leaf-commit,
+// whereas storage tries are committed 'inline'.
+func newCommitter(onleaf LeafCallback) *committer {
+	h := committerPool.Get().(*committer)
+	h.onleaf = onleaf
+	if onleaf != nil {
+		h.leafCh = make(chan *Leaf, LeafChanSize)
+	}
+	return h
+}
+
+func returnCommitterToPool(h *committer) {
+	h.onleaf = nil
+	h.leafCh = nil
+	committerPool.Put(h)
+}
+
+// commitNeeded returns 'false' if the given node is already in sync with db
+func (h *committer) commitNeeded(n node) bool {
+	hash, dirty := n.cache()
+	return hash == nil || dirty
+}
+
+// hash collapses a node down into a hash node, also returning a copy of the
+// original node initialized with the computed hash to replace the original one.
+func (h *committer) commit(n node, db *Database, force bool) (node, error) {
+	// If we're not storing the node, just hashing, use available cached data
+	hash, dirty := n.cache()
+	if hash != nil && !dirty {
+		return hash, nil
+	}
+	if db == nil {
+		return nil, errors.New("no db provided")
+	}
+	// Commit children. then parent
+	// Remove the dirty flag.
+	switch cn := n.(type) {
+	case *shortNode:
+		// Commit child
+		collapsed := cn.copy()
+		if _, ok := cn.Val.(valueNode); !ok {
+			if childV, err := h.commit(cn.Val, db, false); err != nil {
+				return nil, err
+			} else {
+				collapsed.Val = childV
+			}
+		}
+		// The key needs to be copied, since we're delivering it to database
+		collapsed.Key = hexToCompact(cn.Key)
+		hashedNode := h.store(collapsed, db, force, true)
+		if hn, ok := hashedNode.(hashNode); ok {
+			cn.flags.dirty = false
```

Comment:
This is an odd statement. It changes the dirtyness of the original node, which would be a data race (other threads might be reading). Trie nodes are assumed immutable and this changes the internals of an existing thing. This seems very wrong.

Reply:
Does this really differ from the original implementation in hasher.go?

Reply:
Hm, yes, it does